documentation/src/main/docs/kafka/transactions.md

```diff
@@ -0,0 +1,93 @@
+This is used to identify the same producer instance across application restarts.
+By default, it is not configured, and transactions cannot be used.
+When it is configured, the producer is limited to idempotent delivery, therefore `enable.idempotence=true` is implied.
+
+* For __only__ reading transactional messages, a consumer needs to explicitly configure its `isolation.level` property to `read_committed`.
+This will make sure that the consumer will deliver only records committed inside a transaction, and filter out messages from aborted transactions.
+
+* It should also be noted that this does not mean all records atomically written inside a transaction will be read atomically by the consumer.
+Transactional producers can write to multiple topics and partitions inside a transaction, but consumers do not know where the transaction starts or ends.
+Not only multiple consumers inside a consumer group can share those partitions,
+all records which were part of a single transaction can be consumed in different batch of records by a consumer.
+
+
+Kafka connector provides `KafkaTransactions` custom emitter for writing records inside a transaction.
+Before using a transaction we need to make sure that `transactional.id` is configured on the channel properties.
+
+    mp.messaging.outgoing.tx-out-example.transactional.id=example-tx-producer
+
+It can be used as a regular emitter `@Channel`:
+
+``` java
+{{ insert('kafka/outbound/KafkaTransactionalProducer.java') }}
+```
+
+!!!note
+    When `transactional.id` is provided `KafkaProducer#initTransactions` is called when the underlying Kafka producer is created.
+
+The function given to the `withTransaction` method receives a `TransactionalEmitter` for producing records, and returns a `Uni` that provides the result of the transaction.
+If the processing completes successfully, the producer is flushed and the transaction is committed.
+If the processing throws an exception, returns a failing `Uni`, or marks the `TransactionalEmitter` for abort, the transaction is aborted.
+
+If this method is called on a Vert.x context, the processing function is also called on that context.
+Otherwise, it is called on the sending thread of the producer.
+
+!!!important
+    A transaction is considered _in progress_ from the call to the `withTransaction` until the returned `Uni` results in success or failure.
+    While a transaction is in progress subsequent calls to the `withTransaction`, including nested one's inside the given function, will throw `IllegalStateException`.
+    Note that, in Reactive Messaging the execution of processing methods is already serialized, unless `@Blocking(ordered = false)` is used.
+    If `withTransaction` can be called concurrently, for example from a REST endpoint, it is recommended to limit the concurrency of the execution.
+    This can be done using the `@Bulkhead` annotation from Microprofile Fault Tolerance.
```

Comment:
Or by marking the method (which calls `withTransaction`) as `synchronized`?

Reply:
Only making the caller synchronized would not help, as the transaction is ongoing until the `Uni` returns. That's why the `@Bulkhead(1)` works, as it'll schedule another request (if any) once the previous one returns.

Reply:
Ah ah, true! But bulkhead has a limited queue, so it will sooner or later start bailing out. It isn't meant for serializing requests, but for limiting concurrency, so this usage is a little creative :-)

Reply:
I agree, but at least it is configurable. I honestly couldn't think of anything else that will "serialize" requests from imperative code, without being a reactive stream itself.